### PR TITLE
Feature/python support

### PR DESCRIPTION
**Pull request by mbuerki**

### CODE REVIEW

File: cmake/protobuf-generate-cpp.cmake.in [hunk 1]
```diff
@@ -96,6 +96,21 @@ function(PROTOBUF_CATKIN_GENERATE_CPP2 BASE_PATH SRCS HDRS)
   file(MAKE_DIRECTORY ${PROTO_FILE_INSTALL_DIR})
   list(APPEND _protobuf_include_path -I ${PROTO_FILE_INSTALL_DIR})
 
+  set(PYTHON_PROTOC_ARG_VALUE "")
+  set(PYTHON_PROTOC_ARG_FLAGS "")
+  set(PROTOC_PYTHON_COMMENT "")
+  if (${PROTOBUF_COMPILE_PYTHON})
```
Comment (eggerk) [inline cmake/protobuf-generate-cpp.cmake.in]: rm space

File: cmake/protobuf-generate-cpp.cmake.in [hunk 2]
```diff
@@ -96,6 +96,21 @@ function(PROTOBUF_CATKIN_GENERATE_CPP2 BASE_PATH SRCS HDRS)
...
+    # The path of the Python package under devel/lib/python../dist-packages/{project_name}
```
Comment (eggerk) [inline cmake/protobuf-generate-cpp.cmake.in]: .

File: cmake/protobuf-generate-cpp.cmake.in [hunk 3]
```diff
@@ -105,6 +120,12 @@ function(PROTOBUF_CATKIN_GENERATE_CPP2 BASE_PATH SRCS HDRS)
     # BASE_PATH. E.g., for base path "proto" and file 
     # "proto/project/definitions.proto", FIL_REL_DIR would be "project".
     string(REGEX REPLACE "${BASE_PATH}/?" "" FIL_REL_DIR ${FIL_REL_DIR})
+
+    # Variable for the protobuf share folder (e.g., ../devel/share/proto/project-name)
```
Comment (eggerk) [inline cmake/protobuf-generate-cpp.cmake.in]: .

File: cmake/protobuf-generate-cpp.cmake.in [hunk 4]
```diff
@@ -117,25 +138,49 @@ function(PROTOBUF_CATKIN_GENERATE_CPP2 BASE_PATH SRCS HDRS)
...
               "${OUTPUT_BASE_FILE_NAME}.pb.h"
               "${PROTO_GENERATED_HEADERS_INSTALL_DIR}/${FIL_REL_DIR}/${FIL_WE}.pb.h"
       COMMAND ${CMAKE_COMMAND} -E copy
               ${ABS_FIL}
-              "${PROTO_FILE_INSTALL_DIR}/${FIL_REL_DIR}/${FIL_WE}.proto"
+              "${PROTO_SHARE_FOLDER}/${FIL_WE}.proto"
       DEPENDS ${ABS_FIL}
-      COMMENT "Running C++ protocol buffer compiler on ${FIL}"
-      VERBATIM )
+      COMMENT "Running C++${PROTOC_PYTHON_COMMENT} protocol buffer compiler on ${FIL}."
```
Comment (eggerk) [inline cmake/protobuf-generate-cpp.cmake.in]: rm variable and write `(and python if requested)` instead

File: cmake/protobuf-generate-cpp.cmake.in [hunk 5]
```diff
@@ -117,25 +138,49 @@ function(PROTOBUF_CATKIN_GENERATE_CPP2 BASE_PATH SRCS HDRS)
...
   endforeach()
 
   set_source_files_properties(${${SRCS}} ${${HDRS}} PROPERTIES GENERATED TRUE)
   set(${SRCS} ${${SRCS}} PARENT_SCOPE)
-  set(${HDRS} ${${HDRS}} PARENT_SCOPE)
-endfunction()
+  set(${HDRS} ${${HDRS}} PARENT_SCOPE)  
 
+
+  if (${PROTOBUF_COMPILE_PYTHON})
```
Comment (eggerk) [inline cmake/protobuf-generate-cpp.cmake.in]: rm space

File: cmake/protobuf-generate-cpp.cmake.in [hunk 6]
```diff
@@ -117,25 +138,49 @@ function(PROTOBUF_CATKIN_GENERATE_CPP2 BASE_PATH SRCS HDRS)
...
+    # The filepath of the Python package init file.
+    set(PYTHON_INIT_FILEPATH "${PYTHON_PACKAGE_PATH}/__init__.py")
+    # Clear the __init__.py file first.
+    add_custom_target(clear_init ALL
+      COMMAND ${CMAKE_COMMAND} -E remove ${PYTHON_INIT_FILEPATH}
```
Comment (eggerk) [inline cmake/protobuf-generate-cpp.cmake.in]: won't this break python packages?

File: cmake/protobuf-generate-cpp.cmake.in [hunk 7]
```diff
@@ -117,25 +138,49 @@ function(PROTOBUF_CATKIN_GENERATE_CPP2 BASE_PATH SRCS HDRS)
...
+      COMMENT "Cleared and wrote __init__.py."
+      VERBATIM)
+
+    # The Python files must be listed in a __init__.py file.
+    foreach(FIL ${ARGN})
+      get_filename_component(FIL_WE ${FIL} NAME_WE) 
+      # The python filename must not contain '-'. 
+      # They are automatically replacted with '_' by protoc.
```
Comment (eggerk) [inline cmake/protobuf-generate-cpp.cmake.in]: up

File: cmake/protobuf-generate-cpp.cmake.in [hunk 8]
```diff
@@ -117,25 +138,49 @@ function(PROTOBUF_CATKIN_GENERATE_CPP2 BASE_PATH SRCS HDRS)
...
+      string(REGEX REPLACE "-" "_" PYTHON_FILENAME ${FIL_WE})
+      add_custom_target(append_init_${PYTHON_FILENAME} ALL
+        COMMAND ${CMAKE_COMMAND} -E echo "from ${PYTHON_FILENAME}_pb2 import *" >> ${PYTHON_INIT_FILEPATH}
```
Comment (eggerk) [inline cmake/protobuf-generate-cpp.cmake.in]: this doesn't seem very nice
how would you use the proto in python?

File: cmake/protobuf-generate-cpp.cmake.in [hunk 9]
```diff
@@ -117,25 +138,49 @@ function(PROTOBUF_CATKIN_GENERATE_CPP2 BASE_PATH SRCS HDRS)
       OUTPUT "${OUTPUT_BASE_FILE_NAME}.pb.cc"
              "${OUTPUT_BASE_FILE_NAME}.pb.h"
              "${PROTO_GENERATED_HEADERS_INSTALL_DIR}/${FIL_WE}.pb.h"
-             "${PROTO_FILE_INSTALL_DIR}/${PROJECT_NAME}/${FIL_WE}.proto"
+             "${PROTO_SHARE_FOLDER}/${FIL_WE}.proto"
       COMMAND  "${CATKIN_DEVEL_PREFIX}/bin/protoc"
-      ARGS --cpp_out  ${OUTPUT_FOLDER_BASE} ${_protobuf_include_path} ${ABS_FIL}
+      ARGS --cpp_out ${OUTPUT_FOLDER_BASE} ${PYTHON_PROTOC_ARG_FLAG} ${PYTHON_PROTOC_ARG_VALUE} ${_protobuf_include_path} ${ABS_FIL}
```
Comment (eggerk) [inline cmake/protobuf-generate-cpp.cmake.in]: would it make sense to create the proto python in the build folder and then copy it into the develspace?

File: cmake/protobuf-generate-cpp.cmake.in (hunk 8 above)
Comment (mbuerki) [reply cmake/protobuf-generate-cpp.cmake.in]: can you explain why you think this is not nice? 
and what is unclear about how somebody would use the proto in python?

File: cmake/protobuf-generate-cpp.cmake.in (hunk 4 above)
Comment (mbuerki) [reply cmake/protobuf-generate-cpp.cmake.in]: why? this way, if you compile the workspace without python, it'll say "Running C++ protocol buffer..." and if you compile with python, it'll say "Runnig C++ and Python protocol buffer..."... isnt' that nice?

File: cmake/protobuf-generate-cpp.cmake.in (hunk 1 above)
Comment (eggerk) [reply cmake/protobuf-generate-cpp.cmake.in]: don't forget this

File: cmake/protobuf-generate-cpp.cmake.in [hunk 10]
```diff
@@ -96,15 +96,31 @@ function(PROTOBUF_CATKIN_GENERATE_CPP2 BASE_PATH SRCS HDRS)
...
+  endif()
+
   foreach(FIL ${ARGN})
     get_filename_component(ABS_FIL ${FIL} ABSOLUTE)
     get_filename_component(FIL_WE ${FIL} NAME_WE)
     get_filename_component(FIL_REL_DIR ${FIL} DIRECTORY)
 
     # FIL_REL_DIR contains the path to the proto file, starting inside
-    # BASE_PATH. E.g., for base path "proto" and file 
+    # BASE_PATH. E.g., for base path "proto" and file.
```
Comment (eggerk) [inline cmake/protobuf-generate-cpp.cmake.in]: rm . (sentence continues on next line)

File: cmake/protobuf-generate-cpp.cmake.in (hunk 4 above)
Comment (eggerk) [reply cmake/protobuf-generate-cpp.cmake.in]: less complexity :) but it's also ok if you keep it

File: cmake/protobuf-generate-cpp.cmake.in [hunk 11]
```diff
@@ -117,25 +133,39 @@ function(PROTOBUF_CATKIN_GENERATE_CPP2 BASE_PATH SRCS HDRS)
...
+      VERBATIM)
   endforeach()
 
   set_source_files_properties(${${SRCS}} ${${HDRS}} PROPERTIES GENERATED TRUE)
   set(${SRCS} ${${SRCS}} PARENT_SCOPE)
-  set(${HDRS} ${${HDRS}} PARENT_SCOPE)
-endfunction()
+  set(${HDRS} ${${HDRS}} PARENT_SCOPE)  
+
+  if (${PROTOBUF_COMPILE_PYTHON})
```
Comment (eggerk) [inline cmake/protobuf-generate-cpp.cmake.in]: rm space

File: cmake/protobuf-generate-cpp.cmake.in [hunk 12]
```diff
@@ -117,25 +133,39 @@ function(PROTOBUF_CATKIN_GENERATE_CPP2 BASE_PATH SRCS HDRS)
...
+    # The path of the Python package under devel/lib/python../dist-packages/{project_name}.
+    set(PYTHON_PACKAGE_PATH ${CATKIN_DEVEL_PREFIX}/${CATKIN_PACKAGE_PYTHON_DESTINATION})
+    file(MAKE_DIRECTORY ${PYTHON_PACKAGE_PATH})
+    # The filepath of the Python package init file.
+    set(PYTHON_INIT_FILEPATH "${PYTHON_PACKAGE_PATH}/__init__.py")
+    if (NOT EXISTS ${PYTHON_INIT_FILEPATH})
```
Comment (eggerk) [inline cmake/protobuf-generate-cpp.cmake.in]: rm space

File: cmake/protobuf-generate-cpp.cmake.in (hunk 1 above)
Comment (mbuerki) [reply cmake/protobuf-generate-cpp.cmake.in]: urghh, forgot to push the change to remote...

**Pull request merged**